File: python/src/symphony/tracker/memory.py
from symphony.tracker.base import TrackerAdapter
from symphony.models import Issue


class MemoryTracker(TrackerAdapter):
    def __init__(
        self,
        issues: list[Issue] | None = None,
        active_states: list[str] | None = None,
        terminal_states: list[str] | None = None,
    ):
        self._issues = {i.id: i for i in (issues or [])}
        self._active_states = active_states or ["Todo", "In Progress"]
        self._terminal_states = terminal_states or ["Done", "Closed", "Cancelled", "Duplicate"]

    def fetch_candidate_issues(self) -> list[Issue]:
        lower_active = [s.lower() for s in self._active_states]
        return [
            i for i in self._issues.values()
            if i.state.lower() in lower_active
        ]

    def fetch_issues_by_states(self, state_names: list[str]) -> list[Issue]:
        if not state_names:
            return []
        lower = [s.lower() for s in state_names]
        return [i for i in self._issues.values() if i.state.lower() in lower]

    def fetch_issue_states_by_ids(self, issue_ids: list[str]) -> list[Issue]:
        return [self._issues[iid] for iid in issue_ids if iid in self._issues]

File: typescript/src/tracker/memory.ts
import type { Issue } from '../models'
import type { TrackerAdapter } from './base'

export class MemoryTracker implements TrackerAdapter {
  private issues: Map<string, Issue> = new Map()

  constructor(private activeStates: string[] = ['Todo', 'In Progress']) {}

  addIssue(issue: Issue): void {
    this.issues.set(issue.id, issue)
  }

  async fetchCandidateIssues(): Promise<Issue[]> {
    return Array.from(this.issues.values()).filter((i) =>
      this.activeStates.includes(i.state)
    )
  }

  async fetchIssuesByStates(stateNames: string[]): Promise<Issue[]> {
    return Array.from(this.issues.values()).filter((i) =>
      stateNames.includes(i.state)
    )
  }

  async fetchIssueStatesByIds(issueIds: string[]): Promise<Issue[]> {
    return issueIds
      .map((id) => this.issues.get(id))
      .filter((i): i is Issue => i !== undefined)
  }

  async updateIssueState(issueId: string, stateName: string): Promise<void> {
    const issue = this.issues.get(issueId)
    if (!issue) throw new Error(`Issue ${issueId} not found`)
    this.issues.set(issueId, { ...issue, state: stateName })
  }
}
